web/src/main/java/se/inera/intyg/intygstjanst/web/service/impl/PuFilterServiceImpl.java

```diff
@@ -0,0 +1,129 @@
+package se.inera.intyg.intygstjanst.web.service.impl;
+
+import com.google.common.base.Joiner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+import se.inera.intyg.infra.integration.pu.model.PersonSvar;
+import se.inera.intyg.infra.integration.pu.services.PUService;
+import se.inera.intyg.infra.sjukfall.dto.IntygData;
+import se.inera.intyg.intygstjanst.web.service.PuFilterService;
+import se.inera.intyg.schemas.contract.Personnummer;
+
+import java.lang.invoke.MethodHandles;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+public class PuFilterServiceImpl implements PuFilterService {
+
+    private static final String SEKRETESS_SKYDDAD_NAME_PLACEHOLDER = "Skyddad personuppgift";
+    private static final String SEKRETESS_SKYDDAD_NAME_UNKNOWN = "Namn okänt";
+
+    private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass().getSimpleName());
+
+    private final PUService puService;
+
+    public PuFilterServiceImpl(PUService puService) {
+        this.puService = puService;
+    }
+
+    @Override
+    public void enrichWithPatientNameAndFilter(List<IntygData> sickLeaves, boolean filterProtectedPerson) {
+       final var personSvarMap = fetchPersons(sickLeaves);
+
+        final var i = sickLeaves.iterator();
+        while (i.hasNext()) {
+            final var item = i.next();
+
+            final var pnr = getPersonnummerOfOptional(item.getPatientId());
+            if (pnr.isEmpty()) {
+                i.remove();
+                LOG.warn("Problem parsing personnummer returned by PU service. Removing from list of sjukfall.");
+                continue;
+            }
+
+            final var personSvar = personSvarMap.get(pnr.get());
+            final var patientNotFound = personSvar.getStatus() == PersonSvar.Status.NOT_FOUND;
+            if (personSvar.getStatus() == PersonSvar.Status.FOUND || patientNotFound) {
+                if (patientNotFound || personSvar.getPerson().isSekretessmarkering()) {
+
+                    final var updatedName = patientNotFound ? SEKRETESS_SKYDDAD_NAME_UNKNOWN : SEKRETESS_SKYDDAD_NAME_PLACEHOLDER;
+                    item.setPatientNamn(updatedName);
+
+                    if (filterProtectedPerson) {
```

Comment:
The tests you found in Rehabstod needs to be covered here:
- testSekretessmarkeradIsIncludedWhenUserIsLakareAsRehabkoordinatorAndSignedTheIntyg
- testSekretessmarkeradIsExcludedWhenUserIsLakareAsRehabkoordinatorOnSameUnitButIsNotSigning

The rule is that only a doctor should see any protected patient if the doctor was the issuer of the sick leave certificate.

What do you think about a solution like this?
filterProtectedPerson change from boolean -> string. It contains the doctorId that we should allow to show protected patients for. If no doctorId is provided, no protected patients should be displayed/return. If a doctorId is provided it needs to match the doctorId on the IntygData.

filterProtectedPerson could change to something that describes the use a bit better. Don't have a great naming suggestion yet :-)